### PR TITLE
[ENH] Bagging Forecaster

### LINKED ISSUE #2247: [ENH] Bagging Forecaster
We can now use the new bootstrapping transformers (`MovingBlockBootstrap`, `STLBootstrapTransformer`) and the new vectorize feature  to implement a `BaggingForecaster` as described in:

1. Bergmeir, C., Hyndman, R. J., & Benítez, J. M. (2016). Bagging exponential smoothing methods using STL decomposition and Box-Cox transformation. International Journal of Forecasting, 32(2), 303-312
2. Hyndman, R.J., & Athanasopoulos, G. (2021) Forecasting: principles and practice, 3rd edition, OTexts: Melbourne, Australia. OTexts.com/fpp3, Chapter 12.5. Accessed on February 13th 2022.

The point prediction should be the mean of the bootstrapped time series predictions and the distributional forecast should be a different [`tfp.distributions.Empirical`](https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/Empirical) for each time step! 

link to the online version of [2]: https://otexts.com/fpp3/bootstrap.html

### PR DESCRIPTION
**Pull request by ltsaprounis**
#### Reference Issues/PRs

Fixes #2247 

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [ ] I've added unit tests and made sure they pass locally.

##### For new estimators
- [ ] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

### CODE REVIEW

**Comment by fkiraly** [comment]
this is a mysterious issue that is recurring whenever we build composites
previous times, it was solved by writing a proper `_update` function
reason is that the cutoff of the internal estimator is not being updated properly.

But that eases only the symptoms, I have not succeeded finding the cause.
Possibly also related to this: https://github.com/alan-turing-institute/sktime/issues/2118

My hope is that properly cleaning up the `update` family functions will solve this, see https://github.com/alan-turing-institute/sktime/issues/982

**Comment by fkiraly** [comment]
Minimal code to reproduce the issue:

```
from sktime.utils._testing.series import _make_series
from sktime.forecasting.compose import BaggingForecaster
from sktime.forecasting.model_selection import SlidingWindowSplitter

X1 = X[:24]
X2 = X[24:36]
f.fit(X1, fh=[1, 2])

cv = SlidingWindowSplitter(start_with_window=False)
f.update_predict(X2, cv=cv)
```

**Comment by fkiraly** [comment]
hm, that actually works, curious

**Comment by fkiraly** [comment]
This PR should fix the test failure (the symptom) but not the cause:
https://github.com/alan-turing-institute/sktime/pull/2531
why it fixes the test failure: at the start of the test, `fit` is called, so it would prevent #2555 from happening. That prevents the condition under which #2556 materializes.

**Comment by fkiraly** [comment]
and this will fix the cause:
https://github.com/alan-turing-institute/sktime/pull/2557
@ltsaprounis, could you kindly review?

**Comment by ltsaprounis** [comment]
the same tests fail again 😞

**Comment by fkiraly** [comment]
Your tests are failing since you need to implement behaviour in `_update` when `update_params` is false.
I assume you just want to cann your components' update with `update_params=False`?

**Comment by ltsaprounis** [comment]
Implemented an `_update` to see if what happens. I managed to reduce the number of failing tests from 8 to 2, but I can't make sense of the 2 that are failing. Looks like the new y passed in update is not the correct based on the time series frequency, but if that was the case all estimators would fail here.

**Comment by fkiraly** [comment]
@khrapovs, funny stuff is happening here with the `ForecastingHorizon` and `freq`. Any idea??

**Comment by ltsaprounis** [comment]
Thanks for having a look at this @khrapovs. 
Here is a code snippet that reproduces the error:
```python
from sktime.forecasting.compose._bagging import BaggingForecaster
from sktime.transformations.bootstrap import STLBootstrapTransformer
from sktime.utils.estimators import MockForecaster
from sktime.utils._testing.series import _make_series
from sktime.forecasting.model_selection import (
    SlidingWindowSplitter,
    temporal_train_test_split,
)

estimator_instance = BaggingForecaster(STLBootstrapTransformer(sp=3), MockForecaster())
n_columns = 1
fh_int_oos = [2, 5]
window_length = 1
step_length = 5
update_params = False

y = _make_series(n_columns=n_columns, all_positive=True, index_type="datetime")
y_train, y_test = temporal_train_test_split(y)
cv = SlidingWindowSplitter(
    fh_int_oos,
    window_length=window_length,
    step_length=step_length,
    start_with_window=False,
)
estimator_instance.fit(y_train, fh=fh_int_oos)
y_pred = estimator_instance.update_predict(
    y_test, cv=cv, update_params=update_params
)
```

I still can't figure out what's wrong so any help is really appreciated!

**Comment by khrapovs** [comment]
@ltsaprounis If I set the breakpoint here: sktime/datatypes/_utilities.py:106, and start debugging, I see first this:
```
obj.tail(3)
Out[4]: 
2000-02-04    5.033695
2000-02-05    6.121072
2000-02-06    4.013062
Freq: D, dtype: float64
```
Here I can clearly see that the `DatetimeIndex` has the frequency attached to it. In the next iteration I see a completely different object, with a `MultiIndex`:
```
obj.tail(3)
Out[5]: 
                            None
series_id   time_index          
synthetic_9 2000-02-04  4.527963
            2000-02-05  5.294585
            2000-02-06  5.459265
```
Moreover, the last level of this index has no frequency:
```
obj.tail().index.get_level_values(-1)
Out[8]: 
DatetimeIndex(['2000-02-02', '2000-02-03', '2000-02-04', '2000-02-05',
               '2000-02-06'],
              dtype='datetime64[ns]', name='time_index', freq=None)
```
I am not familiar with the estimator in question. Could you perhaps explain, why in the first iteration the index is flat and in the second it has two levels. Maybe you already know why in the second case the index loses the frequency?

**Comment by ltsaprounis** [comment]
@khrapovs thanks for looking into this and sorry for taking ages to reply. 

What you see in the index is expected.
The Bagging Forecast first uses a bootstrapping transformer to generate synthetic time series then forecasts them all independently using the vectorizer utility for hierarchical data and then outputs the mean of the forecasts as the point forecast and the empirical quantiles from the forecasts of the multiple series. 

The frequency correction should be happening in the vectorizer here -> [`sktime.datatypes._vectorize.VectorizedDF.__getitem__`](https://github.com/alan-turing-institute/sktime/blob/c3c0895561cd985dc98d037b7f738c965f83bb7b/sktime/datatypes/_vectorize.py#L152)

I think the problem is somewhere in the composition and the fact that _y is not updated correctly.
I'll keep looking into it.

**Comment by khrapovs** [comment]
Interesting. I have set the breakpoint at `sktime/datatypes/_vectorize.py:227` conditional on `item.index.freq is None`. At some point the object `item` has gaps in it:
```
item.tail()
Out[2]: 
                None
time_index          
2000-02-03  3.208213
2000-02-04  1.589545
2000-02-05  2.475610
2000-02-06  1.886764
2000-02-11  2.621118
```
Notice that there are no observations between 06 and 11. Clearly, `pd.infer_freq(item.index)` is helpless here. Any idea why is there a gap?

**Comment by ltsaprounis** [comment]
> At some point the object `item` has gaps in it:

This is the main problem, yes!
But it looks like `self._y ` (self being the estimator) has this issue for all estimators passed to it!
If that's true a lot more forecasters should be failing here cause the time series is not continuous.

I used `make_mock_estimator` to check what is passed until the failure. 
Check log at the appendix section in the bottom. 

The last logged call is for `_predict_moving_cutoff`.
`_predict_moving_cutoff` basically runs a for loop on the cutoffs from the passed CV splitter object to generate a new y and then calls `update_predict_single` with this new y (named `y_new`). That then calls update using `y_new`. It looks like the problem is that there is a gap between `y_new` and `self._y` so when we combine these 2 to re-fit all hell breaks loose!

## APPENDIX - Mock estimator log

```python

[('clone_tags',
  {'estimator': MockForecaster(),
   'tag_names': ['requires-fh-in-fit', 'enforce_index_type']}),
 ('set_tags',
  {'tag_dict': {'enforce_index_type': None, 'requires-fh-in-fit': False}}),
 ('fit',
  {'X': None,
   'fh': [2, 5],
   'y': 2000-01-01    1.088085
2000-01-02    3.111395
2000-01-03    1.741014
2000-01-04    2.319973
2000-01-05    2.775293
2000-01-06    3.059729
2000-01-07    2.580515
2000-01-08    2.922489
2000-01-09    2.971179
2000-01-10    3.025412
2000-01-11    3.995545
2000-01-12    3.597300
2000-01-13    3.178451
2000-01-14    2.371518
2000-01-15    3.664336
2000-01-16    1.885746
2000-01-17    2.116500
2000-01-18    2.775376
2000-01-19    1.762940
2000-01-20    2.648619
2000-01-21    3.041079
2000-01-22    2.952915
2000-01-23    2.655888
2000-01-24    3.084312
2000-01-25    3.603354
2000-01-26    1.000000
2000-01-27    2.749619
2000-01-28    3.206969
2000-01-29    3.906913
2000-01-30    2.191515
2000-01-31    3.727156
2000-02-01    1.666026
2000-02-02    2.766656
2000-02-03    1.298680
2000-02-04    3.988411
2000-02-05    2.525282
2000-02-06    1.829676
Freq: D, dtype: float64}),
 ('reset', {}),
 ('get_params', {'deep': False}),
 ('clone_tags',
  {'estimator': MockForecaster(),
   'tag_names': ['requires-fh-in-fit', 'enforce_index_type']}),
 ('set_tags',
  {'tag_dict': {'enforce_index_type': None, 'requires-fh-in-fit': False}}),
 ('_check_fh', {'fh': [2, 5]}),
 ('get_tag',
  {'raise_error': True,
   'tag_name': 'requires-fh-in-fit',
   'tag_value_default': None}),
 ('get_tags', {}),
 ('_check_X_y',
  {'X': None,
   'y': 2000-01-01    1.088085
2000-01-02    3.111395
2000-01-03    1.741014
2000-01-04    2.319973
2000-01-05    2.775293
2000-01-06    3.059729
2000-01-07    2.580515
2000-01-08    2.922489
2000-01-09    2.971179
2000-01-10    3.025412
2000-01-11    3.995545
2000-01-12    3.597300
2000-01-13    3.178451
2000-01-14    2.371518
2000-01-15    3.664336
2000-01-16    1.885746
2000-01-17    2.116500
2000-01-18    2.775376
2000-01-19    1.762940
2000-01-20    2.648619
2000-01-21    3.041079
2000-01-22    2.952915
2000-01-23    2.655888
2000-01-24    3.084312
2000-01-25    3.603354
2000-01-26    1.000000
2000-01-27    2.749619
2000-01-28    3.206969
2000-01-29    3.906913
2000-01-30    2.191515
2000-01-31    3.727156
2000-02-01    1.666026
2000-02-02    2.766656
2000-02-03    1.298680
2000-02-04    3.988411
2000-02-05    2.525282
2000-02-06    1.829676
Freq: D, dtype: float64}),
 ('get_tag',
  {'raise_error': True,
   'tag_name': 'y_inner_mtype',
   'tag_value_default': None}),
 ('get_tags', {}),
 ('get_tag',
  {'raise_error': True,
   'tag_name': 'X_inner_mtype',
   'tag_value_default': None}),
 ('get_tags', {}),
 ('get_tag',
  {'raise_error': True, 'tag_name': 'scitype:y', 'tag_value_default': None}),
 ('get_tags', {}),
 ('get_tag',
  {'raise_error': True, 'tag_name': 'scitype:y', 'tag_value_default': None}),
 ('get_tags', {}),
 ('_update_y_X',
  {'X': None,
   'enforce_index_type': None,
   'y': 2000-01-01    1.088085
2000-01-02    3.111395
2000-01-03    1.741014
2000-01-04    2.319973
2000-01-05    2.775293
2000-01-06    3.059729
2000-01-07    2.580515
2000-01-08    2.922489
2000-01-09    2.971179
2000-01-10    3.025412
2000-01-11    3.995545
2000-01-12    3.597300
2000-01-13    3.178451
2000-01-14    2.371518
2000-01-15    3.664336
2000-01-16    1.885746
2000-01-17    2.116500
2000-01-18    2.775376
2000-01-19    1.762940
2000-01-20    2.648619
2000-01-21    3.041079
2000-01-22    2.952915
2000-01-23    2.655888
2000-01-24    3.084312
2000-01-25    3.603354
2000-01-26    1.000000
2000-01-27    2.749619
2000-01-28    3.206969
2000-01-29    3.906913
2000-01-30    2.191515
2000-01-31    3.727156
2000-02-01    1.666026
2000-02-02    2.766656
2000-02-03    1.298680
2000-02-04    3.988411
2000-02-05    2.525282
2000-02-06    1.829676
Freq: D, dtype: float64}),
 ('_set_cutoff_from_y',
  {'y': 2000-01-01    1.088085
2000-01-02    3.111395
2000-01-03    1.741014
2000-01-04    2.319973
2000-01-05    2.775293
2000-01-06    3.059729
2000-01-07    2.580515
2000-01-08    2.922489
2000-01-09    2.971179
2000-01-10    3.025412
2000-01-11    3.995545
2000-01-12    3.597300
2000-01-13    3.178451
2000-01-14    2.371518
2000-01-15    3.664336
2000-01-16    1.885746
2000-01-17    2.116500
2000-01-18    2.775376
2000-01-19    1.762940
2000-01-20    2.648619
2000-01-21    3.041079
2000-01-22    2.952915
2000-01-23    2.655888
2000-01-24    3.084312
2000-01-25    3.603354
2000-01-26    1.000000
2000-01-27    2.749619
2000-01-28    3.206969
2000-01-29    3.906913
2000-01-30    2.191515
2000-01-31    3.727156
2000-02-01    1.666026
2000-02-02    2.766656
2000-02-03    1.298680
2000-02-04    3.988411
2000-02-05    2.525282
2000-02-06    1.829676
Freq: D, dtype: float64}),
 ('_fit',
  {'X': None,
   'fh': ForecastingHorizon([2, 5], dtype='int64', is_relative=True),
   'y': 2000-01-01    1.088085
2000-01-02    3.111395
2000-01-03    1.741014
2000-01-04    2.319973
2000-01-05    2.775293
2000-01-06    3.059729
2000-01-07    2.580515
2000-01-08    2.922489
2000-01-09    2.971179
2000-01-10    3.025412
2000-01-11    3.995545
2000-01-12    3.597300
2000-01-13    3.178451
2000-01-14    2.371518
2000-01-15    3.664336
2000-01-16    1.885746
2000-01-17    2.116500
2000-01-18    2.775376
2000-01-19    1.762940
2000-01-20    2.648619
2000-01-21    3.041079
2000-01-22    2.952915
2000-01-23    2.655888
2000-01-24    3.084312
2000-01-25    3.603354
2000-01-26    1.000000
2000-01-27    2.749619
2000-01-28    3.206969
2000-01-29    3.906913
2000-01-30    2.191515
2000-01-31    3.727156
2000-02-01    1.666026
2000-02-02    2.766656
2000-02-03    1.298680
2000-02-04    3.988411
2000-02-05    2.525282
2000-02-06    1.829676
Freq: D, dtype: float64}),
 ('update_predict',
  {'X': None,
   'cv': SlidingWindowSplitter(fh=[2, 5], start_with_window=False, step_length=5,
           window_length=1),
   'reset_forecaster': True,
   'update_params': False,
   'y': 2000-02-07    2.794208
2000-02-08    2.301699
2000-02-09    1.511276
2000-02-10    2.991909
2000-02-11    2.928313
2000-02-12    3.532948
2000-02-13    3.558503
2000-02-14    2.000523
2000-02-15    2.320293
2000-02-16    2.912753
2000-02-17    1.804444
2000-02-18    1.808553
2000-02-19    1.982570
Freq: D, dtype: float64}),
 ('check_is_fitted', {}),
 ('_check_X_y',
  {'X': None,
   'y': 2000-02-07    2.794208
2000-02-08    2.301699
2000-02-09    1.511276
2000-02-10    2.991909
2000-02-11    2.928313
2000-02-12    3.532948
2000-02-13    3.558503
2000-02-14    2.000523
2000-02-15    2.320293
2000-02-16    2.912753
2000-02-17    1.804444
2000-02-18    1.808553
2000-02-19    1.982570
Freq: D, dtype: float64}),
 ('get_tag',
  {'raise_error': True,
   'tag_name': 'y_inner_mtype',
   'tag_value_default': None}),
 ('get_tags', {}),
 ('get_tag',
  {'raise_error': True,
   'tag_name': 'X_inner_mtype',
   'tag_value_default': None}),
 ('get_tags', {}),
 ('get_tag',
  {'raise_error': True, 'tag_name': 'scitype:y', 'tag_value_default': None}),
 ('get_tags', {}),
 ('get_tag',
  {'raise_error': True, 'tag_name': 'scitype:y', 'tag_value_default': None}),
 ('get_tags', {}),
 ('_predict_moving_cutoff',
  {'X': None,
   'cv': SlidingWindowSplitter(fh=[2, 5], start_with_window=False, step_length=5,
           window_length=1),
   'reset_forecaster': True,
   'update_params': False,
   'y': 2000-02-07    2.794208
2000-02-08    2.301699
2000-02-09    1.511276
2000-02-10    2.991909
2000-02-11    2.928313
2000-02-12    3.532948
2000-02-13    3.558503
2000-02-14    2.000523
2000-02-15    2.320293
2000-02-16    2.912753
2000-02-17    1.804444
2000-02-18    1.808553
2000-02-19    1.982570
Freq: D, dtype: float64})]
```

**Comment by ltsaprounis** [comment]
**OK - I'm formally blaming the test.**

The way the test case is structured there is a gap between the data passed in fit and the data passed in update.
Therefore the reconstructed time series (self._y and y from update) has missing dates and everything breaks. 

To reproduce:
```python
from sktime.utils._testing.series import _make_series
from sktime.forecasting.model_selection import (
    SlidingWindowSplitter,
    temporal_train_test_split,
)

n_columns = 1
fh_int_oos = [2, 5]
window_length = 1
step_length = 5

y = _make_series(n_columns=n_columns, all_positive=True, index_type="datetime")
y_train, y_test = temporal_train_test_split(y)
cv = SlidingWindowSplitter(
    fh_int_oos,
    window_length=window_length,
    step_length=step_length,
    start_with_window=False,
)

print(f"max date seen during fit: {y_train.index.max()}")
counter = 0
for train_idx, test_idx in cv.split(y_test):
    print(f"split {counter} min date seen during update: {y_test.iloc[train_idx].index.min()}")
    counter += 1

# Prints:
# max date seen during fit: 2000-02-06 00:00:00
# split 0 min date seen during update: NaT
# split 1 min date seen during update: 2000-02-11 00:00:00
```

@fkiraly - Would appreciate your take on this.

**Comment by ltsaprounis** [comment]
And in #2701 I'm officially proposing a fix for the test! 🧐

**Comment by fkiraly** [comment]
@ltsaprounis, have you checked whether your fix for the test addresses the failures?

**Comment by ltsaprounis** [comment]
That test passes but I accidentally committed an earlier version of #2686 that I used to debug things and it breaks the make_mock_estimator tests 🤦‍♂️. I'll update it when #2686 gets merged.

File: sktime/forecasting/compose/_bagging.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+from sktime.forecasting.base import BaseForecaster
+from sktime.transformations.base import BaseTransformer
+from sktime.transformations.bootstrap import STLBootstrapTransformer
+from sktime.utils.estimators import MockForecaster
+
+
+class BaggingForecaster(BaseForecaster):
+    """Bagged "Bootstrap Aggregating" Forecasts.
+
+    Bagged Forecasts are obtained by forecasting bootstrapped time series and then
```
Comment (fkiraly) [inline sktime/forecasting/compose/_bagging.py:27]: I think this section should give the reader enough information to tell what the algorithm is, precisely.

E.g., what exactly is bootstrapped, how? Is there a paper reference? If yes, which algorithm is it, from the paper reference?

File: sktime/forecasting/compose/_bagging.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+    ----------
+    bootstrapping_transformer : BaseTransformer
+        Bootrstrapping Transformer that takes a series as input and returns a panel
+        of bootstrapped time series
+    forecaster : BaseForecaster
+        A valid sktime Forecaster
+
+    See Also
+    --------
+    sktime.transformations.bootstrap.MovingBlockBootstrapTransformer :
```
Comment (fkiraly) [inline sktime/forecasting/compose/_bagging.py:46]: should one of these be the default for the `bootstrapping_transformer` arg?

File: sktime/forecasting/compose/_bagging.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+    are aggregated using the mean function [1]. Prediction intervals and quantiles are
+    calculated for each time point in the forecasting horizon by calculating the sampled
+    forecast quantiles.
+
+    Bergmeir et al. (2016) [2] show that, on average, bagging ETS forecasts gives better
+    forecasts than just applying ETS directly.
+
+    Parameters
+    ----------
+    bootstrapping_transformer : BaseTransformer
```
Comment (fkiraly) [inline sktime/forecasting/compose/_bagging.py:38]: would `bootstrap_transformer` be a better name?

File: sktime/forecasting/compose/_bagging.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+        Bootrstrapping Transformer that takes a series as input and returns a panel
```
Comment (fkiraly) [inline sktime/forecasting/compose/_bagging.py:39]: typo

File: sktime/forecasting/compose/_bagging.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+        of bootstrapped time series
+    forecaster : BaseForecaster
+        A valid sktime Forecaster
+
+    See Also
+    --------
+    sktime.transformations.bootstrap.MovingBlockBootstrapTransformer :
+        Transofrmer that applies the Moving Block Bootstrapping method to create
```
Comment (fkiraly) [inline sktime/forecasting/compose/_bagging.py:47]: typo

**Review by fkiraly — CHANGES_REQUESTED** [review]
Looks good!

I added some comments, some about docs, some about naming and param defaults.

The logic itself looks fine.

Other question: why does this need to be univariate? "computing the mean" is not something that is difficult to generalize...

**Comment by ltsaprounis** [comment]
@fkiraly - ready for review! Sorry for the delay. 
I added new default estimators for as in the paper referenced for Bagging ETS forecasts as suggested in the comments.
I agree that it could in principle be extended to multivariate, but I'd like to leave that for later! This is basically an implementation of the methodology explained in Forecasting Principles and Practice book and I don't want to deviate from it too much for now:
- [Hyndman, R.J., & Athanasopoulos, G. (2021) Forecasting: principles and practice, 3rd edition, OTexts: Melbourne, Australia. OTexts.com/fpp3, Chapter 12.5. Accessed on February 13th 2022.](https://otexts.com/fpp3/bootstrap.html)

File: sktime/forecasting/compose/_bagging.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+        update_params : bool, optional (default=True)
+            whether model parameters should be updated
+
+        Returns
+        -------
+        self : reference to self
+        """
+        # Need to construct a completely new y out of ol self._y and y and then
+        # fit_treansform the transformer and re-fit the foreaster.
+        _y = deepcopy(self._y)
```
Comment (fkiraly) [inline sktime/forecasting/compose/_bagging.py:272]: why create this deepcopy? None of the estimators should have side effects (and this is being tested)

File: sktime/forecasting/compose/_bagging.py [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+    intervals and quantiles are calculated for each time point in the forecasting
+    horizon by calculating the sampled forecast quantiles.
+
+    Bergmeir et al. (2016) [2] show that, on average, bagging ETS forecasts gives better
+    forecasts than just applying ETS directly. The default bootstraping transformer
+    and forecaster are selected as in [2].
+
+    Parameters
+    ----------
+    bootstrap_transformer : BaseTransformer (default=None)
```
Comment (fkiraly) [inline sktime/forecasting/compose/_bagging.py:39]: default should say the transformer instead - the function has arg `None`, but it "means" `STLBootstrapTransformer`

File: sktime/forecasting/compose/_bagging.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+        Bootstrapping Transformer that takes a series as input  and returns a panel
+        of bootstrapped time series if not specified
+        sktime.transformations.bootstrap.STLBootstrapTransformer is used.
+    forecaster : BaseForecaster (default=None)
```
Comment (fkiraly) [inline sktime/forecasting/compose/_bagging.py:43]: same

File: sktime/forecasting/compose/_bagging.py [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+        Returns
+        -------
+        params : dict or list of dict, default = {}
+            Parameters to create testing instances of the class
+            Each dict are parameters to construct an "interesting" test instance, i.e.,
+            `MyClass(**params)` or `MyClass(**params[i])` creates a valid test instance.
+            `create_test_instance` uses the first (or only) dictionary in `params`
+        """
+        params = [
+            {},
```
Comment (fkiraly) [inline sktime/forecasting/compose/_bagging.py:294]: should have a second param set where non-defaults are being used for the components

**Review by fkiraly — CHANGES_REQUESTED** [review]
Really nice!

I have one blocking comment: we need to test this also with non-defaults substituted for the components.

Other, non-blocking comments are above:
* `deepcopy` in `_update`, really needed?
* should `_update` instead call the transformer's  `update`?
* docstring clarity

File: sktime/forecasting/compose/_bagging.py [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+    forecast quantiles.
+
+    Bergmeir et al. (2016) [2] show that, on average, bagging ETS forecasts gives better
+    forecasts than just applying ETS directly.
+
+    Parameters
+    ----------
+    bootstrapping_transformer : BaseTransformer
+        Bootrstrapping Transformer that takes a series as input and returns a panel
+        of bootstrapped time series
```
[the remark this answers is not in the document]
Comment (fkiraly) [reply sktime/forecasting/compose/_bagging.py]: still valid comment

File: sktime/forecasting/compose/_bagging.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fkiraly) [reply sktime/forecasting/compose/_bagging.py]: still valid comment

File: sktime/forecasting/compose/_bagging.py [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+from sktime.transformations.base import BaseTransformer
+from sktime.transformations.bootstrap import STLBootstrapTransformer
+
+
+class BaggingForecaster(BaseForecaster):
+    """Forecast a time series by aggregating forecasts from its bootstraps.
+
+    Bagged "Bootstrap Aggregating" Forecasts are obtained by forecasting bootstrapped
+    time series and then aggregating the resulting forecasts. For the point forecast,
+    the different forecasts are aggregated using the mean function [1]. Prediction
```
Comment (fkiraly) [inline sktime/forecasting/compose/_bagging.py:29]: users might appreciate having a choice here, e.g., an `agg_fun` param. Though we do not need to add it in this PR.

File: sktime/forecasting/compose/_bagging.py (hunk 7 above)
Comment (ltsaprounis) [reply sktime/forecasting/compose/_bagging.py]: It was fixed last time as well, the code snippet here is Outdated.

**Review by fkiraly — APPROVED** [review]
Alright, we're testing other transformers now, so the blocking issue is addressed. Good to go!

Feel free to fix the other ones or not, and feel free to merge once tests pass.

**Comment by ltsaprounis** [comment]
Deep-copy is now gone. 
I implemented _update just because of the composites issue, I think it's safer to practically re-fit here instead.

**Comment by fkiraly** [comment]
>I implemented _update just because of the composites issue

Sounds like deep trauma...

**Pull request merged**